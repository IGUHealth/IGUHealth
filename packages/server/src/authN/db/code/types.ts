import * as s from "zapatos/schema";

export type AuthorizationCode = s.authorization_code.JSONSelectable & {
  is_expired: boolean;
};
