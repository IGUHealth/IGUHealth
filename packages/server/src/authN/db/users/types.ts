import * as s from "zapatos/schema";
export const USER_QUERY_COLS = <const>[
  "id",
  "tenant",
  "scope",
  "email",
  "first_name",
  "last_name",
  "email_verified",
  "role",
  "fhir_user_id",
  "fhir_user_versionid",
];

export type User = s.users.OnlyCols<typeof USER_QUERY_COLS>;

export type LoginParameters = {
  "email-password" : {
    email: string;
    password: string;
  },
  "oidc-provider": {
    email: string;
    provider: string;
  }
};
