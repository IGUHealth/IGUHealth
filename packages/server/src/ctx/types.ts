import type { Logger } from "pino";

import {
  ResourceType,
  AResource,
  CapabilityStatement,
  canonical,
  uri,
} from "@iguhealth/fhir-types/r4/types";
import { FHIRClientAsync } from "@iguhealth/client/interface";

import { Lock } from "../synchronization/interfaces.js";
import { IOCache } from "../cache/interface.js";
import { TerminologyProvider } from "../terminology/interface.js";
import { EncryptionProvider } from "../encryption/provider/interface.js";

declare const __workspace: unique symbol;
export type Workspace = string & { [__workspace]: string };

declare const __author: unique symbol;
export type Author = string & { [__author]: string };

export interface FHIRServerCTX {
  encryptionProvider?: EncryptionProvider;
  inTransaction?: boolean;
  workspace: Workspace;
  author: Author;
  terminologyProvider: TerminologyProvider;
  // Services setup
  logger: Logger<unknown>;
  capabilities: CapabilityStatement;
  cache: IOCache<FHIRServerCTX>;
  client: FHIRClientAsync<FHIRServerCTX>;
  lock: Lock<unknown>;
  user_access_token?: string;
  resolveTypeToCanonical: (type: uri) => canonical | undefined;
  resolveCanonical: <T extends ResourceType>(
    type: T,
    url: string
  ) => AResource<T> | undefined;
}
