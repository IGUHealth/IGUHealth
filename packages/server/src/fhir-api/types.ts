import type { RouterParamContext } from "@koa/router";
import type Koa from "koa";
import type { Logger } from "pino";
import * as db from "zapatos/db";
import * as s from "zapatos/schema";

import { FHIRClientAsync } from "@iguhealth/client/interface";
import {
  AccessPolicy,
  ClientApplication,
  Membership,
  OperationDefinition,
  canonical,
  code,
  id,
  uri,
} from "@iguhealth/fhir-types/r4/types";
import {
  AllResourceTypes,
  FHIR_VERSION,
  Resource,
} from "@iguhealth/fhir-types/versions";
import {
  AccessToken,
  AccessTokenPayload,
  CUSTOM_CLAIMS,
  IGUHEALTH_ISSUER,
  Subject,
  TenantId,
} from "@iguhealth/jwt";

import { AuthorizationCodeManagement } from "../authN/db/code/interface.js";
import { UserManagement } from "../authN/db/users/interface.js";
import { User } from "../authN/db/users/types.js";
import { OIDCRouteHandler } from "../authN/oidc/index.js";
import { sessionLogin, sessionLogout } from "../authN/oidc/session/index.js";
import type { IOCache } from "../cache/interface.js";
import { EmailProvider } from "../email/interface.js";
import type { EncryptionProvider } from "../encryption/provider/interface.js";
import type { TerminologyProvider } from "../fhir-terminology/interface.js";
import type { Lock } from "../synchronization/interfaces.js";

type MakeOptional<T, K extends keyof T> = Omit<T, K> & {
  [P in K]?: T[P];
};

export namespace KoaExtensions {
  export type KoaIGUHealthContext = Koa.ParameterizedContext<
    KoaExtensions.IGUHealth,
    RouterParamContext<KoaExtensions.IGUHealth, KoaIGUHealthContext>
  >;

  export type OIDC = {
    oidc: {
      sessionLogin: typeof sessionLogin;
      sessionLogout: typeof sessionLogout;
      isAuthenticated: (
        ctx: Parameters<OIDCRouteHandler>[0],
      ) => Promise<boolean>;

      user?: User;
      tenant: TenantId;
      userManagement: UserManagement;
      codeManagement: AuthorizationCodeManagement;
      client?: ClientApplication;
      parameters: {
        state?: string;
        responseType?: code;
        response_type?: string;
        client_id?: string;
        redirect_uri?: string;
        scope?: string;
      };
    };
  };

  export type IGUHealth = OIDC &
    IGUHealthServices & {
      iguhealth: MakeOptional<IGUHealthServerCTX, "user">;
    };

  export type IGUHealthServices = {
    allowSignup?: boolean;
    __user__?: AccessTokenPayload<s.user_role>;
    __access_token__?: AccessToken<s.user_role>;
    iguhealth: MakeOptional<IGUHealthServerCTX, "user" | "tenant">;
  };

  /**
   * Verifies whether ctx is IGUHealthServerCTX with user.
   * @param ctx CTX that should be verified as user associated
   * @returns  Verifies whether ctx is IGUHealthServerCTX with user
   */
  export function isFHIRServerAuthorizedUserCTX(
    ctx: IGUHealthServerCTX | Omit<IGUHealthServerCTX, "user">,
  ): ctx is IGUHealthServerCTX {
    return (ctx as IGUHealthServerCTX).user !== undefined;
  }
}

export interface UserContext {
  payload: AccessTokenPayload<s.user_role>;
  resource?: Membership | ClientApplication | OperationDefinition | null;
  accessPolicies?: AccessPolicy[];
  accessToken?: string;
}

export interface IGUHealthServerCTX {
  // Server Information
  tenant: TenantId;
  user: UserContext;

  // FHIR Client
  client: FHIRClientAsync<IGUHealthServerCTX>;

  // Services
  db: db.Queryable;
  logger: Logger<string>;
  lock: Lock<unknown>;
  cache: IOCache<IGUHealthServerCTX>;
  terminologyProvider?: TerminologyProvider;
  encryptionProvider?: EncryptionProvider;
  emailProvider?: EmailProvider;

  // Utilities
  resolveTypeToCanonical: (
    fhirVersion: FHIR_VERSION,
    type: uri,
  ) => canonical | undefined;
  resolveCanonical: <
    FHIRVersion extends FHIR_VERSION,
    Type extends AllResourceTypes,
  >(
    fhirVersion: FHIRVersion,
    type: Type,
    url: canonical,
  ) => Resource<FHIRVersion, Type> | undefined;
}

export function rootJWT(tenant: TenantId): AccessTokenPayload<s.user_role> {
  return {
    iss: IGUHEALTH_ISSUER,
    [CUSTOM_CLAIMS.RESOURCE_TYPE]: "Membership",
    [CUSTOM_CLAIMS.RESOURCE_ID]: "system" as id,
    [CUSTOM_CLAIMS.ROLE]: "admin",
    [CUSTOM_CLAIMS.TENANT]: tenant,
    sub: "system" as Subject,
  };
}

/**
 * For certain operations like parameter retrievals must treat context as a system user.
 * To bypass any access control checks, we can use this function to create a new context.
 * @param ctx The current context
 * @returns A new context with the user set to system.
 */
export function asRoot(
  ctx: Omit<IGUHealthServerCTX, "user">,
): IGUHealthServerCTX {
  return {
    ...ctx,
    tenant: ctx.tenant,
    user: {
      payload: rootJWT(ctx.tenant),
    },
  };
}
