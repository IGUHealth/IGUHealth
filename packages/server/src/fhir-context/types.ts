import type { Logger } from "pino";
import * as db from "zapatos/db";
import * as s from "zapatos/schema";

import { FHIRClientAsync } from "@iguhealth/client/interface";
import {
  AResource,
  AccessPolicy,
  CapabilityStatement,
  ClientApplication,
  Membership,
  OperationDefinition,
  ResourceType,
  canonical,
  code,
  uri,
} from "@iguhealth/fhir-types/r4/types";

import { AuthorizationCodeManagement } from "../authN/db/code/interface.js";
import { UserManagement } from "../authN/db/users/interface.js";
import { IGUHEALTH_ISSUER, JWT } from "../authN/token.js";
import type { IOCache } from "../cache/interface.js";
import { EmailProvider } from "../email/interface.js";
import type { EncryptionProvider } from "../encryption/provider/interface.js";
import type { TerminologyProvider } from "../fhir-terminology/interface.js";
import type { Lock } from "../synchronization/interfaces.js";

export namespace KoaContext {
  export type OIDC = {
    oidc: {
      userManagement: UserManagement;
      codeManagement: AuthorizationCodeManagement;
      context: { scope: "global" } | { scope: "tenant"; tenant: TenantId };
      client?: ClientApplication;
      parameters: {
        state?: string;
        responseType?: code;
        response_type?: string;
        client_id?: string;
        redirect_uri?: string;
        scope?: string;
      };
    };
  };

  export type FHIR<C> = C &
    OIDC &
    FHIRServices & {
      FHIRContext: Omit<FHIRServerCTX, "user"> | FHIRServerCTX;
    };

  export type FHIRServices = {
    emailProvider?: EmailProvider;
    postgres: db.Queryable;
    FHIRContext: Omit<FHIRServerCTX, "user" | "tenant">;
  };

  /**
   * Verifies whether ctx is FHIRServerCTX with user.
   * @param ctx CTX that should be verified as user associated
   * @returns  Verifies whether ctx is FHIRServerCTX with user
   */
  export function isFHIRServerAuthorizedUserCTX(
    ctx: FHIRServerCTX | Omit<FHIRServerCTX, "user">,
  ): ctx is FHIRServerCTX {
    return (ctx as FHIRServerCTX).user !== undefined;
  }
}

declare const __tenant: unique symbol;
export type TenantId = string & { [__tenant]: string };
export interface TenantClaim {
  id: TenantId;
  userRole: s.user_role;
}

export interface UserContext {
  role: s.user_role;
  jwt: JWT;
  resource?: Membership | ClientApplication | OperationDefinition | null;
  accessPolicies?: AccessPolicy[];
  accessToken?: string;
}

export interface FHIRServerCTX {
  // Server Information
  capabilities: CapabilityStatement;
  tenant: TenantId;
  user: UserContext;

  // FHIR Client
  client: FHIRClientAsync<FHIRServerCTX>;

  // Services
  db: db.Queryable;
  logger: Logger<string>;
  lock: Lock<unknown>;
  cache: IOCache<FHIRServerCTX>;
  terminologyProvider: TerminologyProvider;
  encryptionProvider?: EncryptionProvider;

  // Contextual Information.
  // If this is set to true, then the current request is part of a transaction.
  inTransaction?: boolean;

  // Utilities
  resolveTypeToCanonical: (type: uri) => canonical | undefined;
  resolveCanonical: <T extends ResourceType>(
    type: T,
    url: string,
  ) => AResource<T> | undefined;
}

/**
 * For certain operations like parameter retrievals must treat context as a system user.
 * To bypass any access control checks, we can use this function to create a new context.
 * @param ctx The current context
 * @returns A new context with the user set to system.
 */
export function asSystemCTX(ctx: Omit<FHIRServerCTX, "user">): FHIRServerCTX {
  return {
    ...ctx,
    tenant: ctx.tenant,
    user: {
      role: "admin",
      jwt: {
        iss: IGUHEALTH_ISSUER,
        sub: "system",
      } as JWT,
    },
  };
}
