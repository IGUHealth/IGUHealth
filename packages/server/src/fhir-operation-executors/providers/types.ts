import { FHIRRequest } from "@iguhealth/client/lib/types";
import { id } from "@iguhealth/fhir-types/lib/generated/r4/types";
import { FHIR_VERSION, R4, ResourceType } from "@iguhealth/fhir-types/versions";

import { IGUHealthServerCTX } from "../../fhir-server/types.js";

export type Payload<I> = {
  ctx: {
    SEC_TOKEN: string;
    API_URL: string;
    tenant: IGUHealthServerCTX["tenant"];
    level: FHIRRequest<FHIR_VERSION>["level"];
    resourceType?: ResourceType<R4>;
    id?: id;
  };
  input: I;
};
