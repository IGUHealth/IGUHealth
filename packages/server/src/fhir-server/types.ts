import type { RouterParamContext } from "@koa/router";
import type Koa from "koa";
import type { Logger } from "pino";
import * as s from "zapatos/schema";

import { FHIRClientAsync } from "@iguhealth/client/interface";
import {
  AccessPolicyV2,
  ClientApplication,
  IdentityProvider,
  Membership,
  OperationDefinition,
  canonical,
  code,
  id,
} from "@iguhealth/fhir-types/r4/types";
import {
  FHIR_VERSION,
  Resource,
  ResourceType,
} from "@iguhealth/fhir-types/versions";
import {
  AccessToken,
  AccessTokenPayload,
  CUSTOM_CLAIMS,
  JWT,
  Subject,
  TenantId,
} from "@iguhealth/jwt/types";

import { getIssuer } from "../authN/oidc/constants.js";
import { SYSTEM_APP } from "../authN/oidc/hardcodedClients/system-app.js";
import { OIDCRouteHandler } from "../authN/oidc/index.js";
import { Scope } from "../authN/oidc/scopes/parse.js";
import {
  sessionCredentialsLogin,
  sessionLogout,
} from "../authN/oidc/session/index.js";
import type { IOCache } from "../cache/interface.js";
import { EmailProvider } from "../email/interface.js";
import type { EncryptionProvider } from "../encryption/provider/interface.js";
import type { TerminologyProvider } from "../fhir-terminology/interface.js";
import {
  IQueue,
  IQueueBatch,
} from "../queue/implementations/providers/interface.js";
import { SearchEngine } from "../search-stores/interface.js";
import { User } from "../storage/interfaces/authAdmin/authAdmin.js";
import { PostgresStore } from "../storage/postgres/index.js";
import type { LockProvider } from "../synchronization/interfaces.js";

type MakeOptional<T, K extends keyof T> = Omit<T, K> & {
  [P in K]?: T[P];
};

export namespace KoaExtensions {
  export type KoaIGUHealthContext = Koa.ParameterizedContext<
    KoaExtensions.IGUHealth,
    RouterParamContext<KoaExtensions.IGUHealth, KoaIGUHealthContext>
  >;

  export type OIDC = {
    oidc: {
      sessionLogin: typeof sessionCredentialsLogin;
      sessionLogout: typeof sessionLogout;
      isAuthenticated: (
        ctx: Parameters<OIDCRouteHandler>[0],
      ) => Promise<boolean>;
      user?: User;
      client?: ClientApplication;
      identityProviders?: IdentityProvider[];
      parameters: {
        client_id?: string;
        client_secret?: string;

        code_verifier?: string;
        code_challenge_method?: string;
        code_challenge?: string;
        state?: string;
        responseType?: code;
        response_type?: string;
        redirect_uri?: string;
        scope?: string;
      };
      launch?: Record<string, id>;
      scopes?: Scope[];
    };
  };

  export type IGUHealth = OIDC &
    IGUHealthServices & {
      iguhealth: MakeOptional<IGUHealthServerCTX, "user">;
    };

  export type IGUHealthServices = {
    allowSignup?: boolean;
    __user__: AccessTokenPayload<s.user_role>;
    __access_token__: AccessToken<s.user_role>;
    corsNonce: string;
    iguhealth: MakeOptional<IGUHealthServerCTX, "user" | "tenant">;
  };

  /**
   * Verifies whether ctx is IGUHealthServerCTX with user.
   * @param ctx CTX that should be verified as user associated
   * @returns  Verifies whether ctx is IGUHealthServerCTX with user
   */
  export function isFHIRServerAuthorizedUserCTX(
    ctx: IGUHealthServerCTX | Omit<IGUHealthServerCTX, "user">,
  ): ctx is IGUHealthServerCTX {
    return (ctx as IGUHealthServerCTX).user !== undefined;
  }
}

export interface UserContext {
  payload: AccessTokenPayload<s.user_role>;
  accessToken?: JWT<AccessTokenPayload<s.user_role>>;
  resource: Membership | ClientApplication | OperationDefinition;
  accessPolicies?: AccessPolicyV2[];
  scope?: Scope[];
}

export interface IGUHealthServerCTX {
  environment: string;
  // Server Information
  tenant: TenantId;
  user: UserContext;

  queue: IQueue | IQueueBatch;

  // FHIR Client
  client: FHIRClientAsync<IGUHealthServerCTX>;

  // Storage
  search: SearchEngine<IGUHealthServerCTX>;
  store: PostgresStore<IGUHealthServerCTX>;

  // Services
  cache?: IOCache<Pick<IGUHealthServerCTX, "tenant">>;
  lock: LockProvider<Pick<IGUHealthServerCTX, "store">>;

  logger: Logger<string>;
  terminologyProvider?: TerminologyProvider;
  encryptionProvider?: EncryptionProvider;
  emailProvider?: EmailProvider;

  // Utilities
  resolveCanonical: <
    Version extends FHIR_VERSION,
    Type extends ResourceType<Version>,
  >(
    ctx: this,
    fhirVersion: Version,
    type: Type,
    url: canonical[],
  ) => Promise<Resource<Version, Type>[]>;
}

function createRootClaims(
  tenant: TenantId,
  clientApp: ClientApplication,
): AccessTokenPayload<s.user_role> {
  return {
    scope: "system/*.*",
    iss: getIssuer(tenant),
    sub: clientApp.id as string as Subject,
    aud: clientApp.id as id,
    [CUSTOM_CLAIMS.RESOURCE_ID]: clientApp.id as id,
    [CUSTOM_CLAIMS.RESOURCE_TYPE]: clientApp.resourceType,
    [CUSTOM_CLAIMS.ROLE]: "owner",
    [CUSTOM_CLAIMS.TENANT]: tenant,
    [CUSTOM_CLAIMS.ACCESS_POLICY_VERSION_IDS]: [],
  };
}

/**
 * For certain operations like parameter retrievals must treat context as a system user.
 * To bypass any access control checks, we can use this function to create a new context.
 * @param ctx The current context
 * @returns A new context with the user set to system.
 */
export function asRoot(
  ctx: Omit<IGUHealthServerCTX, "user">,
): IGUHealthServerCTX {
  const rootClaims = createRootClaims(ctx.tenant, SYSTEM_APP);

  return {
    ...ctx,
    tenant: ctx.tenant,
    user: {
      resource: SYSTEM_APP,
      payload: rootClaims,
    },
  };
}
