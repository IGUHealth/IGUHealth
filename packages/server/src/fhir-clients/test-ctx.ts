import { pino } from "pino";
import { fileURLToPath } from "url";
import * as s from "zapatos/schema";

import { loadArtifacts } from "@iguhealth/artifacts";
import { canonical } from "@iguhealth/fhir-types/r4/types";
import {
  FHIR_VERSION,
  R4,
  R4B,
  Resource,
  ResourceType,
} from "@iguhealth/fhir-types/versions";
import {
  AccessTokenPayload,
  CUSTOM_CLAIMS,
  TenantId,
} from "@iguhealth/jwt/types";

import { IOCache } from "../cache/interface.js";
import getConfigProvider from "../config/index.js";
import { IGUHealthServerCTX } from "../fhir-server/types.js";
import { TerminologyProvider } from "../fhir-terminology/index.js";
import { createSearchStore } from "../search-stores/index.js";
import createStore from "../storage/index.js";
import PostgresLock from "../synchronization/postgres.lock.js";
import { Memory } from "./clients/memory/async.js";

const sds = loadArtifacts({
  fhirVersion: R4,
  loadDevelopmentPackages: true,
  resourceType: "StructureDefinition",
  currentDirectory: fileURLToPath(import.meta.url),
  silence: true,
});

const parameters = loadArtifacts({
  fhirVersion: R4,
  resourceType: "SearchParameter",
  // Don't want to load other searchparameters which could conflict with base for now.
  onlyPackages: [
    "@iguhealth/hl7.fhir.r4.core",
    "@iguhealth/iguhealth.fhir.r4.core",
  ],
  currentDirectory: fileURLToPath(import.meta.url),
  silence: true,
});

class TestCache<CTX extends { tenant: TenantId }> implements IOCache<CTX> {
  _date: Record<string, unknown>;
  constructor() {
    this._date = {};
  }
  get(_ctx: CTX, key: string): Promise<string | number | null> {
    return Promise.resolve((this._date[key] as string | number) || null);
  }
  set(_ctx: CTX, key: string, value: string | number): Promise<void> {
    this._date[key] = value;
    return Promise.resolve();
  }
}

const config = getConfigProvider();
const store = await createStore(config);

export const testServices: IGUHealthServerCTX = {
  config,
  tenant: "tenant" as TenantId,
  store,
  search: await createSearchStore(config),
  lock: new PostgresLock(store.getClient()),
  // @ts-ignore
  user: {
    payload: {
      iss: "test",
      sub: "test-user",
      [CUSTOM_CLAIMS.ROLE]: "member",
    } as AccessTokenPayload<s.user_role>,
  },
  terminologyProvider: new TerminologyProvider(),
  logger: pino<string>(),
  client: new Memory({ [R4]: {}, [R4B]: {} }),
  cache: new TestCache(),
  resolveCanonical: async <
    Version extends FHIR_VERSION,
    Type extends ResourceType<Version>,
    URL extends canonical[],
  >(
    _ctx: IGUHealthServerCTX,
    version: Version,
    type: Type,
    url: URL,
  ): Promise<Resource<Version, Type>[]> => {
    switch (type) {
      case "StructureDefinition": {
        return sds.filter((sd) =>
          url.includes(sd.url as canonical),
        ) as Resource<Version, Type>[];
      }
      case "SearchParameter": {
        return parameters.filter((param) =>
          url.includes(param.url as canonical),
        ) as Resource<Version, Type>[];
      }
      default: {
        return [];
      }
    }
  },
};
