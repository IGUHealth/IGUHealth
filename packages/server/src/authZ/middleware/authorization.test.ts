import { expect, test } from "@jest/globals";

import { FHIRRequest<FHIR_VERSION>, FHIRResponse } from "@iguhealth/client/lib/types";
import {
  AccessPolicyV2,
  Patient,
  code,
  id,
} from "@iguhealth/fhir-types/lib/generated/r4/types";

import { IGUHealthServerCTX } from "../../fhir-server/types.js";
import { testServices } from "../../fhir-clients/test-ctx.js";
import createAuthorizationMiddleware from "./authorization.js";

const authorizationMiddleware = createAuthorizationMiddleware();

test("Authorization test for read access on resource based on type and method", async () => {
  const ctx: IGUHealthServerCTX = {
    ...testServices,
    tenant: testServices.tenant,
    user: {
      ...testServices.user,
      accessPolicies: [
        {
          id: "test-policy" as id,
          resourceType: "AccessPolicyV2",
          engine: "rule-engine" as code,
          name: "test-policy",
          code: "test-policy" as code,
          rule: [
            {
              name: "test-rule",
              target: {
                expression: {
                  expression: "true",
                  language: "text/fhirpath" as code,
                },
              },
              condition: {
                expression: {
                  expression: "%request.type = 'read-request'",
                  language: "text/fhirpath" as code,
                },
              },
            },
          ],
        } as AccessPolicyV2,
      ],
    },
  };

  const responder: ReturnType<typeof createAuthorizationMiddleware> = async (
    context,
  ) => {
    return {
      ...context,
      response: {
        fhirVersion: context.request.fhirVersion,
        type: "read-response",
        level: "instance",
        resource: "Patient",
        id: "1" as id,
        body: {
          resourceType: "Patient",
          id: "1",
        } as Patient,
      } as FHIRResponse,
    };
  };

  expect(
    (
      await authorizationMiddleware(
        {
          ctx: ctx,
          state: {},
          request: {
            type: "read-request",
            level: "instance",
            resource: "Patient",
            id: "1",
          } as FHIRRequest<FHIR_VERSION>,
        },
        responder,
      )
    ).response,
  ).toEqual({
    type: "read-response",
    level: "instance",
    resource: "Patient",
    id: "1" as id,
    body: {
      resourceType: "Patient",
      id: "1",
    } as Patient,
  });

  expect(async () => {
    try {
      await authorizationMiddleware(
        {
          ctx: ctx,
          state: {},
          request: {
            type: "search-request",
            level: "type",
            resource: "Patient",
          } as FHIRRequest<FHIR_VERSION>,
        },
        responder,
      );
    } catch (e) {
      // @ts-ignore
      throw e.operationOutcome;
    }
  }).rejects.toMatchSnapshot();
});
