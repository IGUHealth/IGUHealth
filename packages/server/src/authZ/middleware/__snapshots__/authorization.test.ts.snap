// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authorization test for read access on resource based on type and method 1`] = `
{
  "issue": [
    {
      "code": "forbidden",
      "diagnostics": "No policy has granted access to your request",
      "expression": undefined,
      "severity": "error",
    },
    {
      "code": "forbidden",
      "diagnostics": "Access Denied.",
      "expression": [
        "/rule/0",
      ],
      "extension": [
        {
          "url": "https://iguhealth.app/AccessPolicy/policy-extension",
          "valueReference": {
            "reference": "AccessPolicyV2/test-policy",
          },
        },
      ],
      "severity": "error",
    },
  ],
  "resourceType": "OperationOutcome",
}
`;
